Update node release based on maplibre-js release workflow

acalcutt opened the pull request:
This PR updates the node release workflow so it no longer increments the version on it's own. This should allow it to work with the approval requirements on this repo. I've made this work more like the maplibre-js workflow at https://github.com/maplibre/maplibre-gl-js/blob/b7b86c1018842a139d4ced21a071ca4a3d9f24b5/.github/workflows/release.yml . 

To increment the version now, someone would make a PR that changes the version in package.json located in the root directory (which is used for node). Once that change is made, the workflow automatically checks against the published version on npm. If the version has changed, the binaries get built and published to github using NODE_PRE_GYP_GUTHUB, then the release is published to npm

I've also cherry picked some [workflow fixes](https://github.com/maplibre/maplibre-native/pull/1941) I made in the opengl-2 branch. 
- update '@acalcutt/node-pre-gyp' to support node 20
- put '--fallback-to-build=false' in package.json. I had changed this a while back but it should be false since we don't have GYP build files
- fix 'runner.arch' that should be 'matrix.arch'
- update readme to reflect binaries being built
- remove unused node-ci-mac.yml

I was also thinking about versioning. Maybe in the 'opengl-2' branch we can continue a 5.x release , but in main we can start with '6.0.0-pre.0' (once we get something together for a metal pre-release)


**Testing**
I have created a test release (minus macos support) at
https://github.com/WifiDB/maplibre-native/actions/runs/7368227110/job/20052326170
https://www.npmjs.com/package/@acalcutt/maplibre-gl-native-test/v/5.2.1-pre.9
https://github.com/WifiDB/maplibre-native/releases/tag/node-v5.2.1-pre.9

**Wishlist**
A way to update release changelog based on https://github.com/maplibre/maplibre-native/blob/main/platform/node/CHANGELOG.md

**Notes**
After this is approved I plan to cherry pick these changes over to  'opengl-2' so I could also make a updated release with macos support there, so a squash and merge may be easier if this is merged.

acalcutt commented:
Any suggestions on how to fix the 'pre-commit.ci' `run:` error

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -1,53 +1,44 @@
-          git config user.email github-actions@github.com
-          git commit -a -m "Update node version to ${{ env.NPM_VERSION }} (${{ github.event.inputs.version }})"
-          git push
+          latestNpmPackageVersion=$( npm view @maplibre/maplibre-gl-native versions --json | jq '.[-1]' -r )
```
```suggestion
          latestNpmPackageVersion="$( npm view @maplibre/maplibre-gl-native versions --json | jq '.[-1]' -r )"
```

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -1,53 +1,44 @@
-          git commit -a -m "Update node version to ${{ env.NPM_VERSION }} (${{ github.event.inputs.version }})"
-          git push
+          latestNpmPackageVersion=$( npm view @maplibre/maplibre-gl-native versions --json | jq '.[-1]' -r )
+          currentVersion=$( node -e "console.log(require('./package.json').version)" )
```
```suggestion
          currentVersion="$( node -e "console.log(require('./package.json').version)" )"
```

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -273,24 +259,34 @@ jobs:
+      - name: Prepare release
+        id: prepare_release
+        run: |
+          RELEASE_TYPE=$(node -e "console.log(require('semver').prerelease('${{ steps.package-version.outputs.current-version }}') ? 'prerelease' : 'regular')")
```
```suggestion
          RELEASE_TYPE="$(node -e "console.log(require('semver').prerelease('${{ steps.package-version.outputs.current-version }}') ? 'prerelease' : 'regular')")"
```

louwers commented:
@acalcutt You can run

```
pre-commit run actionlint
```

locally or install all pre-commit hooks with

```
pre-commit install
```

(after installing pre-commit with pip or another package manager).

I left some suggestions that will solve the issue I think. `$(command)` needs to be quoted to prevent word splitting (same as with variables).

louwers reviewed (approved):
Looks great!

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -1,53 +1,44 @@
+          latestNpmPackageVersion="$( npm view @maplibre/maplibre-gl-native versions --json | jq '.[-1]' -r )"
+          currentVersion="$( node -e "console.log(require('./package.json').version)" )"
+          if [ "$latestNpmPackageVersion" == "$currentVersion" ]; then
+            echo "changed=false" >> $GITHUB_OUTPUT
```
```suggestion
            echo "changed=false" >> "$GITHUB_OUTPUT"
```

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -1,53 +1,44 @@
+          if [ "$latestNpmPackageVersion" == "$currentVersion" ]; then
+            echo "changed=false" >> $GITHUB_OUTPUT
+          else
+            echo "changed=true" >> $GITHUB_OUTPUT
```
```suggestion
            echo "changed=true" >> "$GITHUB_OUTPUT"
```

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -273,24 +259,34 @@ jobs:
+        id: prepare_release
+        run: |
+          RELEASE_TYPE="$(node -e "console.log(require('semver').prerelease('${{ steps.package-version.outputs.current-version }}') ? 'prerelease' : 'regular')")"
+          echo "release_type=$RELEASE_TYPE" >> $GITHUB_OUTPUT
```
```suggestion
          echo "release_type=$RELEASE_TYPE" >> "$GITHUB_OUTPUT"
```

acalcutt commented:
I changed this a little bit so instead of checking if the current version matched the latest version, it just checks if the version is published.

In testing I realized If I published a newer version on a different branch (like main), it would break publishing of lower version releases (like I want to do with the opengl-2 branch).  Just checking if it is published will allow simultaneous releases and allow it to only publish when needed still.

acalcutt commented:
i still haven't figured out that last pre-commit.ci error, it still doesn't like line 27 but it seems like most of it is in quotes now...

louwers commented on .github/workflows/node-release.yml:
```diff
@@ -1,53 +1,41 @@
-          git commit -a -m "Update node version to ${{ env.NPM_VERSION }} (${{ github.event.inputs.version }})"
-          git push
+          currentVersion="$( node -e "console.log(require('./package.json').version)" )"
+          isPublished="$( npm view @maplibre/maplibre-gl-native versions --json | jq -c --arg cv $currentVersion 'any(. == $cv)' )"
```
I ran `pre-commit run actionlint --all-files` to confirm, this is the last thing!

```suggestion
          isPublished="$( npm view @maplibre/maplibre-gl-native versions --json | jq -c --arg cv "$currentVersion" 'any(. == "$cv")' )"
```

acalcutt replied:
It doesn't work when I quote the "$cv" (I think because it is a variable passed from jq?), but just quoting the "$currentVersion" seemed to make the pre-commit ci happy and this check actually still works.

Skipped because published
https://github.com/WifiDB/maplibre-native/actions/runs/7391761598

New release publish
https://github.com/WifiDB/maplibre-native/actions/runs/7391779968

acalcutt commented:
@louwers I was wondering what you think about this.

In a test branch I added code that adds the node changelog to these releases. This is also based on the maplibre-js release workflow. It takes the version change information from 'platform/node/CHANGELOG.md' and Updates the release with the text for the version posted.
https://github.com/WifiDB/maplibre-native/blob/node_pr_workflow_test/.github/workflows/node-release.yml#L271-L304

This needs this script i modified from maplibre-js, like this one
https://github.com/WifiDB/maplibre-native/blob/node_pr_workflow_test/scripts/release-notes-node.js

I also had to fix the indentation #'s in  'platform/node/CHANGELOG.md'  to match what maplibre-js used
https://raw.githubusercontent.com/WifiDB/maplibre-native/node_pr_workflow_test/platform/node/CHANGELOG.md


Do you think these changes are an improvement? should I merge them in here?

louwers commented:
> Do you think these changes are an improvement? should I merge them in here?

Sure, let's try it out. Maybe in a different PR?

Can you put the script in `platform/node/scripts` (since it is Node.js specific)?

Can you error when no release notes are found?

louwers commented:
@acalcutt I am using this awk one-liner now for iOS:

https://github.com/maplibre/maplibre-native/blob/fe728ebdd8b58edc1d01ce538fcec8dbec247031/.github/workflows/ios-ci.yml#L220

acalcutt commented:
I am going to close this in favor of https://github.com/maplibre/maplibre-native/pull/1995 . I have moved that PR to use awk like recommended
